core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1080,9 +1080,9 @@ class UnifiedLog(@volatile var logStartOffset: Long,
     (updatedProducers, completedTxns.toList, None)
   }
 
-  private def batchMissingRequiredVerification(batch: MutableRecordBatch, requestVerificationGuard: Object): Boolean = {
+  private def batchMissingRequiredVerification(batch: MutableRecordBatch, requestVerificationGuard: VerificationGuard): Boolean = {
     producerStateManager.producerStateManagerConfig().transactionVerificationEnabled() && !batch.isControlBatch &&
-      (requestVerificationGuard != verificationGuard(batch.producerId) || requestVerificationGuard == null)
+      (requestVerificationGuard == null || !requestVerificationGuard.equals(verificationGuard(batch.producerId)))
```

Comment:
What's the reason to change this logic?  Reference comparison works well, handles nulls properly, and we don't have to rely on the autoincremented int other than for diagnostics.

Reply:
I recall part of the conversation referring to the likelihood of collisions. I see that the reference is also 64 bits. 

I can change it back to reference, but since we have the autoincrementing value, I thought it could be useful for readability of the code since object references being used to check equality aren't as common of a pattern in Java.

Reply:
It's not possible to have a collision on comparison of references of 2 separately allocated live objects (unless there is a bug in JVM), there could be collision on the hash codes, but not references.   And it's not just because reference is 64 bits, the references of 2 live objects are never the same, this is guaranteed by JVM.   It's theoretically possible to get a collision on integers (they are just auto-incremented, after all, nobody tracks that lifetime of an object with one id doesn't overlap with lifetime with object of collided id), though it would require a very long time for a 64 bit integer (shouldn't happen in practice in lifetime of a broker).

We definitely have examples of this in Kafka codebase https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java#L164 

This one is written by me :-), but I'm pretty sure I've seen this pattern somewhere else in Kafka.

Reply:
Both implementations are correct. Personally I'd go with Justine's approach since the guard implements equals/hash code/toString. Static code checkers and humans would typically be surprised at reference quality being used in such cases.

Reply:
The second question is: can we avoid the null check altogether by using a sentinel?

Reply:
The sentinel could be value==0. I think that would be a lot cleaner.

Reply:
I think you'd actually want to have a method in `VerificationGuard` to do the comparison, not `equals`. That method would then check for the sentinel and do the right thing. Does that make sense?